larq_compute_engine/tests/strip_lcedequantize_test.py

```diff
@@ -0,0 +1,78 @@
+import sys
+
+import larq as lq
+import pytest
+import tensorflow as tf
+
+from larq_compute_engine.mlir.python.converter import convert_keras_model
+from larq_compute_engine.mlir.python.util import strip_lcedequantize_ops
+
+
+def toy_model_sign(**kwargs):
+    img = tf.keras.layers.Input(shape=(224, 224, 3))
+    x = lq.layers.QuantConv2D(
+        256,
+        kernel_size=3,
+        strides=1,
+        padding="same",
+        pad_values=1,
+        input_quantizer="ste_sign",
+        kernel_quantizer="ste_sign",
+        kernel_constraint="weight_clip",
+    )(img)
+    x = lq.quantizers.SteSign()(x)
+    return tf.keras.Model(inputs=img, outputs=x)
+
+
+def quant(x):
+    return tf.quantization.fake_quant_with_min_max_vars(x, -3.0, 3.0)
+
+
+def toy_model_int8_sign(**kwargs):
+    img = tf.keras.layers.Input(shape=(224, 224, 3))
+    x = quant(img)
+    x = lq.layers.QuantConv2D(
+        256,
+        kernel_size=3,
+        strides=1,
+        padding="same",
+        pad_values=1,
+        input_quantizer="ste_sign",
+        kernel_quantizer="ste_sign",
+        kernel_constraint="weight_clip",
+    )(img)
+    x = lq.quantizers.SteSign()(x)
+    x = quant(x)
+    return tf.keras.Model(inputs=img, outputs=x)
+
+
+@pytest.mark.parametrize("model_cls", [toy_model_sign, toy_model_int8_sign])
+@pytest.mark.parametrize("inference_input_type", [tf.int8, tf.float32])
+@pytest.mark.parametrize("inference_output_type", [tf.int8, tf.float32])
+def test_strip_lcedequantize_ops(
+    model_cls, inference_input_type, inference_output_type
+):
+    model_lce = convert_keras_model(
+        model_cls(),
+        inference_input_type=inference_input_type,
+        inference_output_type=inference_output_type,
+        experimental_default_int8_range=(-6.0, 6.0)
+        if model_cls == toy_model_sign
+        else None,
+        experimental_enable_bitpacked_activations=True,
+    )
+    model_lce = strip_lcedequantize_ops(model_lce)
+    interpreter = tf.lite.Interpreter(model_content=model_lce)
+    input_details = interpreter.get_input_details()
+    assert len(input_details) == 1
+    assert input_details[0]["dtype"] == inference_input_type.as_numpy_dtype
+    output_details = interpreter.get_output_details()
+    assert len(output_details) == 1
+    if inference_output_type == tf.float32:
+        assert output_details[0]["dtype"] == tf.int32.as_numpy_dtype
+    else:
+        assert output_details[0]["dtype"] == inference_output_type.as_numpy_dtype
```

Comment:
Does this mean if output inference type is `tf.int8` the dequantize op is not removed?
@simonmaurer Is this intended behaviour?

Reply:
yes, this is intended behavior.

as discussed here:
> Sorry for the late reply.
> 
> > In that case these checks are not necessary, or am I missing something here?
> 
> To be honest, I am not 100% sure about this. This code was originally adapted from [tensorflow/tensorflow@ec94fab](https://github.com/tensorflow/tensorflow/commit/ec94fabb838b0934eeaf46f1fd22c36b2b3ec333) but it looks like there've been [additional changes in the latest version](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/python/util.py) which would be great to apply here as well in another PR.
> I would need to take a closer look if these errors would ever be thrown here, but if you think they won't, fell free to remove the checks for them in `strip_lcedequantize_ops`.
> 
> Could you still add a unittest similar to [this one](https://github.com/larq/compute-engine/blob/9297ff48c7bec562dc25aaad4853e65a351b710c/larq_compute_engine/tests/end2end_test.py#L203-L219) that tests that the functionality works as expected?
> I do understand that you manually verified that the code works, but it is important for us to have these checks also run on CI so we can ensure that it also works with older versions of TensorFlow and crucially that the code will keep working and we don't accidentally break it in a future PR.

and when looking at the graph of `tf.float32/tf.int8` models there are only `LceDequantize` ops when the model has `tf.float32` outputs.
in other words the `LceDequantize` ops are removed (resulting in the desired tf.int32 outputs) iff the model has `tf.float32` outputs

Reply:
actually it shouldn't differ..as long as there are `LceDequantize` ops as outputs they should be removed to get `tf.int32` outputs, no matter what `tf.float32/tf.int8` ops are inside the model

Reply:
MLIR failed again for the assertion tests

Reply:
You might also want to apply similar changes as done in #635 to this PR, but I don't think this will fix CI either. Looks like the dequantize node is not correctly removed in your example.